core/src/main/scala/kafka/cluster/Replica.scala

```diff
@@ -29,7 +29,7 @@ class Replica(val brokerId: Int,
               val topicPartition: TopicPartition,
               time: Time = Time.SYSTEM,
               initialHighWatermarkValue: Long = 0L,
-              val log: Option[Log] = None) extends Logging {
+              var log: Option[Log] = None) extends Logging {
```

Comment:
It needs to be var because `Partition.maybeDeleteAndSwapFutureReplica()` needs to do `replica.log = futureReplica.log`. Alternative we can replace the current replica with the future replica in Partition.allReplicasMap. But that means we need to copy states such as lastFetchTimeMs from the current Replica to the future Replica. And it takes extra care to maintain this when we add new state in the Replica in the future. Thus I choose the current approach because it is simpler.

Reply:
@junrao Thanks for catching this. I have updated it patch to fix it.